src/ctx.rs

```diff
@@ -0,0 +1,172 @@
+//! This module provided types for context representation could be used in context-sensitive parsing.
+//! See [ctx attribute](attributes/index.html#ctx) for more information.
+
+use core::ops::{Deref, DerefMut};
+use core::str::FromStr;
+
+/// An endian.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Endian {
+    Little,
+    Big,
+}
+
+/// An error returned when parsing a `Endian` using [`from_str`] fails
+///
+/// [`from_str`]: enum.Endian.html#method.from_str
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ParseEndianError {}
+
+impl Endian {
+    /// As same as [`Endian::default`], but const.
+    ///
+    /// [`Endian::default`]: #method.default
+    pub const fn new() -> Self {
+        #[cfg(target_endian = "little")]
+        let endian = Endian::Little;
+
+        #[cfg(target_endian = "big")]
+        let endian = Endian::Big;
+
+        endian
+    }
+
+    /// Is it a little endian.
+    pub fn is_le(&self) -> bool {
+        match self {
+            Endian::Little => true,
+            _ => false,
+        }
+    }
+
+    /// Is it a big endian.
+    pub fn is_be(&self) -> bool {
+        match self {
+            Endian::Big => true,
+            _ => false,
+        }
+    }
+}
+
+impl Default for Endian {
+    /// Return the endianness of the target's CPU.
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FromStr for Endian {
+    type Err = ParseEndianError;
+
+    /// Parse a `Endian` from a string.
+    /// # Examples
+    /// ```rust
+    /// use std::str::FromStr;
+    /// use deku::ctx::Endian;
+    /// assert_eq!(FromStr::from_str("little"), Ok(Endian::Little));
```

Comment:
Any reason to write like this instead of:
```rust
    /// assert_eq!(Endian::from_str("little"), Ok(Endian::Little));
    /// assert_eq!(Endian::from_str("big"), Ok(Endian::Big));
    /// assert!(Endian::from_str("not an endian").is_err());
```

Reply:
Just one unimportant reason: `FromStr::from_str` is more right since in Rust you can write:
```rust
struct A();
impl A {
    fn from_str() {}
}
impl FromStr for A {
    fn from_str(_: &str) -> ... {}
}
let a = A::from_str("a string"); // compile error. `A::from_str` exactly ref to `fn from_str() {}`.
let a: A = FromStr::from_str("a string"); // good
```

Reply:
Ah interesting! I'm surprised there's no clippy lint which would catch this...

I would normally write `Endian::from_str` for clarity, but I see your point